lib/services/access_list.go

```diff
@@ -243,20 +257,166 @@ func IsAccessListOwner(identity tlsca.Identity, accessList *accesslist.AccessLis
 // AccessListMembershipChecker will check if users are members of an access list and
 // makes sure the user is not locked and meets membership requirements.
 type AccessListMembershipChecker struct {
-	members AccessListMemberGetter
+	members AccessListRecursiveMembersGetter
 	locks   LockGetter
 	clock   clockwork.Clock
 }
 
 // NewAccessListMembershipChecker will create a new access list membership checker.
-func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListMemberGetter, locks LockGetter) *AccessListMembershipChecker {
+func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListRecursiveMembersGetter, locks LockGetter) *AccessListMembershipChecker {
 	return &AccessListMembershipChecker{
 		members: members,
 		locks:   locks,
 		clock:   clock,
 	}
 }
 
+func recurseAccessLists(username string, initialList string,
+	lists func(string) ([]string, error),
+	listMember func(username, list string) error,
+	onAcl func(current string) error) error {
+	seen := map[string]struct{}{}
+	queue := []string{initialList}
+
+	for len(queue) > 0 {
+
+		size := len(queue)
+		for i := 0; i < size; i++ {
+			pal := queue[0]
+			queue = queue[1:]
+
+			err := listMember(username, pal)
+			if err != nil && !trace.IsNotFound(err) {
+				return trace.Wrap(err)
+			}
+			if trace.IsNotFound(err) {
+				subAccessListMembers, err := lists(pal)
+				if err != nil {
+					return trace.NotFound("error find access list %s", pal)
+				}
+				for _, next := range subAccessListMembers {
+					if _, ok := seen[next]; ok {
+						return trace.BadParameter("cycles are not allowed between dynamic access list members")
```

Comment:
What about diamond hierarchy where there is no any cycles or other [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph)  ?: 
![Screenshot 2024-06-27 at 14 58 50](https://github.com/gravitational/teleport/assets/22402974/fb7e13e0-a612-4fe2-846e-2b44d69338dc)